### DISCUSSION TITLE
uhppoted-mqtt disconnection

### DISCUSSION THREAD

**Discussion by liviom** [uhppoted/uhppoted#32] [Q&A] [answered]
Dear All

I have the following errors 

`o@UbuntuOH-Data:/opt/uhppoted/uhppoted_v0.8.5$ sudo ./uhppoted-mqtt --console --config /etc/uhppoted/uhppoted.conf  --console
[sudo] password di livio:
2023/06/29 02:14:51 INFO                uhppoted-mqtt service v0.8.x - Linux (PID 357369)
2023/06/29 02:14:51 INFO                START
2023/06/29 02:14:51 WARN   auth         open /var/uhppoted/mqtt.hotp.counters: no such file or directory
2023/06/29 02:14:51 WARN   auth         open /var/uhppoted/mqtt.nonce: no such file or directory
2023/06/29 02:14:51 WARN   auth         open /var/uhppoted/mqtt.nonce.counters: no such file or directory
2023/06/29 02:14:51 WARN                open /etc/uhppoted/mqtt/cards: no such file or directory
2023/06/29 02:14:51 INFO   mqttd        listening on 192.168.0.90:60001
2023/06/29 02:14:51 INFO   mqttd        publishing events to uhppoted/gateway/events
2023/06/29 02:14:51 INFO   listen       Fetching unretrieved events for device ID 423133701
2023/06/29 02:14:51 INFO   mqttd        connected to tcp://192.168.0.90:1883
2023/06/29 02:14:51 INFO   mqttd        subscribed to uhppoted/gateway/requests
2023/06/29 02:14:51 [store]    memorystore del: message 1 not found
2023/06/29 02:14:51 INFO   listen       No unretrieved events for device ID 423133701
2023/06/29 02:14:51 INFO   listen       Initialising event listener
2023/06/29 02:14:51 INFO   listen       Listening
2023/06/29 02:14:56 INFO   watchdog     OK
2023/06/29 02:15:01 INFO   watchdog     OK
2023/06/29 02:15:01 [client]   Connect comms goroutine - error triggered EOF
2023/06/29 02:15:01 [client]   BUG BUG BUG reconnection function is nil <nil>
2023/06/29 02:15:01 ERROR  mqttd        connection to MQTT broker lost (EOF)
2023/06/29 02:15:01 INFO   mqttd        DISCONNECT 1 of 10 in 5m0s
2023/06/29 02:15:06 INFO   watchdog     OK
2023/06/29 02:15:06 INFO   health-check refreshing interface IP address list
2023/06/29 02:15:09 INFO   health-check OK
2023/06/29 02:15:09 WARN   monitoring   no connection to MQTT broker
2023/06/29 02:15:11 INFO   watchdog     OK
2023/06/29 02:15:11 INFO   mqttd        retrying connection to MQTT broker tcp://192.168.0.90:1883
2023/06/29 02:15:11 INFO   mqttd        connected to tcp://192.168.0.90:1883
2023/06/29 02:15:11 INFO   mqttd        subscribed to uhppoted/gateway/requests
2023/06/29 02:15:11 [store]    memorystore del: message 2 not found
2023/06/29 02:15:16 INFO   watchdog     OK
2023/06/29 02:15:21 INFO   watchdog     OK
2023/06/29 02:15:21 [client]   Connect comms goroutine - error triggered EOF
2023/06/29 02:15:21 [client]   BUG BUG BUG reconnection function is nil <nil>
2023/06/29 02:15:21 ERROR  mqttd        connection to MQTT broker lost (EOF)
2023/06/29 02:15:21 INFO   mqttd        DISCONNECT 2 of 10 in 5m0s
2023/06/29 02:15:24 INFO   health-check OK
2023/06/29 02:15:24 WARN   monitoring   no connection to MQTT broker
2023/06/29 02:15:26 INFO   watchdog     OK
2023/06/29 02:15:31 INFO   watchdog     OK
2023/06/29 02:15:31 INFO   mqttd        retrying connection to MQTT broker tcp://192.168.0.90:1883
2023/06/29 02:15:31 INFO   mqttd        connected to tcp://192.168.0.90:1883
2023/06/29 02:15:31 INFO   mqttd        subscribed to uhppoted/gateway/requests
2023/06/29 02:15:31 [store]    memorystore del: message 3 not found
2023/06/29 02:15:36 INFO   watchdog     OK
2023/06/29 02:15:39 INFO   health-check OK
2023/06/29 02:15:41 INFO   watchdog     OK
2023/06/29 02:15:41 [client]   Connect comms goroutine - error triggered EOF
2023/06/29 02:15:41 [client]   BUG BUG BUG reconnection function is nil <nil>
2023/06/29 02:15:41 ERROR  mqttd        connection to MQTT broker lost (EOF)
2023/06/29 02:15:41 INFO   mqttd        DISCONNECT 3 of 10 in 5m0s
2023/06/29 02:15:46 INFO   watchdog     OK
2023/06/29 02:15:51 INFO   watchdog     OK
2023/06/29 02:15:51 INFO   mqttd        retrying connection to MQTT broker tcp://192.168.0.90:1883
2023/06/29 02:15:51 INFO   mqttd        connected to tcp://192.168.0.90:1883
2023/06/29 02:15:51 INFO   mqttd        subscribed to uhppoted/gateway/requests
2023/06/29 02:15:51 [store]    memorystore del: message 4 not found
2023/06/29 02:15:54 INFO   health-check OK
2023/06/29 02:15:56 INFO   watchdog     OK
2023/06/29 02:16:01 INFO   watchdog     OK
2023/06/29 02:16:01 [client]   Connect comms goroutine - error triggered EOF
2023/06/29 02:16:01 [client]   BUG BUG BUG reconnection function is nil <nil>
2023/06/29 02:16:01 ERROR  mqttd        connection to MQTT broker lost (EOF)
2023/06/29 02:16:01 INFO   mqttd        DISCONNECT 4 of 10 in 5m0s
^C2023/06/29 02:16:04 INFO                interrupt
2023/06/29 02:16:04 INFO   mqttd        closing connection to tcp://192.168.0.90:1883
2023/06/29 02:16:04 [client]   status is already disconnected
2023/06/29 02:16:04 INFO   mqttd        closed connection to tcp://192.168.0.90:1883
2023/06/29 02:16:04 INFO                exit
livio@UbuntuOH-Data:/opt/uhppoted/uhppoted_v0.8.5$

`

and after 10 disconnection it stops the app.
this happens also if it is executed out of the console with command "systemctl start uhppoted-mqtt.service"

how can I solve?

thanks in advance for your help

**Comment by uhppoted** [comment]
Hmmm, this seems like it might be a bug in the _Eclipse Paho MQTT_ client.

_client.go [lines 547-552](https://github.com/eclipse/paho.mqtt.golang/blob/aa0a8ad044fe531bbf7336aa6b7e1c9a5031cddf/client.go#L547-L552)_:
```
		reConnDone, err := disDone(true)
		if err != nil {
			ERROR.Println(CLI, "failure whilst reporting completion of disconnect", err)
		} else if reConnDone == nil { // Should never happen
			ERROR.Println(CLI, "BUG BUG BUG reconnection function is nil", err)
		}
```

It looks like there was a fix for another bug in Paho that may be related but it was more than 6 months ago so also not very likely. I'll have to dig into it - can you give me a little more information:

- Are you still using the _mosquito_ broker?
- Did it start occurring after you upgraded either _mosquito_ or _uhppoted-mqtt_ ?
- Does the problem still occur with an earlier version of _uhppoted-mqtt_ ?

**Comment by uhppoted** [comment]
Raised [Issue #12: uhppoted-mqtt disconnection](https://github.com/uhppoted/uhppoted-mqtt/issues/12) in _uhppoted-mqtt_ to track progress.

**Comment by uhppoted** [comment]
Can you also check your _mosquito_ logs to see if there is any related information there? The Paho developer seems to think it might be either a broker issue or possibly invalid topics:

https://github.com/eclipse/paho.mqtt.golang/issues/457

(and if you could attach your _uhppoted.conf_ file that would also be helpful)

**Comment by uhppoted** [comment]
Just a thought - it's been about a year since you set up that system so if you have TLS enabled it might be worth checking to see that the certificates haven't expired.

**Reply by liviom** [reply]
dear
it happens also with previuos version 0.8.0

I don't use TLS, 

mi config files , mqtt secion is the following:

`# MQTT
; mqtt.server.ID = uhppoted
mqtt.connection.broker = tcp://192.168.0.90:1883
mqtt.connection.client.ID = uhppoted_mqttd
; mqtt.connection.username = 
; mqtt.connection.password = 
; mqtt.connection.broker.certificate = /etc/uhppoted/mqtt/broker.cert
; mqtt.connection.client.certificate = /etc/uhppoted/mqtt/client.cert
; mqtt.connection.client.key = /etc/uhppoted/mqtt/client.key
mqtt.topic.root = uhppoted/gateway
mqtt.topic.requests = ./requests
mqtt.topic.replies = ./replies
mqtt.topic.events = ./events
mqtt.topic.system = ./system
; mqtt.alerts.qos = 2
mqtt.alerts.retained = true
mqtt.events.key = events
mqtt.system.key = system
; mqtt.events.index.filepath = /var/uhppoted/mqtt.events.retrieved
; mqtt.permissions.enabled = false
; mqtt.permissions.users = /etc/uhppoted/mqtt.permissions.users
; mqtt.permissions.groups = /etc/uhppoted/mqtt.permissions.groups
; mqtt.cards = /etc/uhppoted/mqtt/cards
mqtt.security.HMAC.required = false
; mqtt.security.HMAC.key = 
mqtt.security.authentication = NONE
; mqtt.security.hotp.range = 8
; mqtt.security.hotp.secrets = /etc/uhppoted/mqtt.hotp.secrets
; mqtt.security.hotp.counters = /var/uhppoted/mqtt.hotp.counters
; mqtt.security.rsa.keys = /etc/uhppoted/mqtt/rsa
mqtt.security.nonce.required = false
mqtt.security.nonce.server = /var/uhppoted/mqtt.nonce
mqtt.security.nonce.clients = /var/uhppoted/mqtt.nonce.counters
mqtt.security.outgoing.sign = false
mqtt.security.outgoing.encrypt = false`

and this is the entire config file
[uhppoted.conf.txt](https://github.com/uhppoted/uhppoted/files/11902331/uhppoted.conf.txt)

yes I use Mosquitto dockerized (with portainer) and in mosquitto log there are no issue while in the console is reporting the disconnections;
I upgrade mosquitto container regularly on every release
Let me know if you need other information.

**Comment by liviom** [comment, marked as answer]
dear 

Solved, It was my fault, there was another instance of uhppoted-mqtt running on another virtual machine (used as failover backup)

stopping the other instance it execute without error and disconnections.

sorry again.

**Comment by uhppoted** [comment]
No worries, I've done that too!! More than once actually :-).

In fact it's the whole raison d'être for the logic around the _10 retries and then shutdown_. Without the delayed retry logic you get an absolute storm of reconnects as each system kicks the other system off the broker and if you happen not to be watching the system at the time the logs can fill the disk in minutes.

Glad you found the problem and thanks for letting me know!
